File: src/main/java/net/minestom/server/statistic/StatisticTypeImpl.java
package net.minestom.server.statistic;

import net.kyori.adventure.key.Key;
import net.minestom.server.registry.BuiltinRegistries;
import net.minestom.server.registry.Registry;
import net.minestom.server.registry.RegistryData;
import net.minestom.server.registry.RegistryKey;

@SuppressWarnings("removal")
record StatisticTypeImpl(Key key, int id) implements StatisticType {
    static final Registry<StatisticType> REGISTRY = RegistryData.createStaticRegistry(BuiltinRegistries.CUSTOM_STAT,
            (namespace, properties) -> new StatisticTypeImpl(Key.key(namespace), properties.getInt("id")));

    static StatisticType get(RegistryKey<StatisticType> key) {
        return REGISTRY.get(key);
    }

    @Override
    public String toString() {
        return name();
    }
}

File: src/main/java/net/minestom/server/statistic/StatisticTypeImpl.kt
package net.minestom.server.statistic

import net.minestom.server.registry.ProtocolObject
import net.minestom.server.registry.Registry
import net.minestom.server.statistic.StatisticTypes
import net.minestom.server.statistic.StatisticType
import net.minestom.server.statistic.StatisticTypeImpl
import net.minestom.server.utils.NamespaceID
import net.minestom.server.statistic.StatisticCategory

internal class StatisticTypeImpl : StatisticType {
    override fun toString(): String {
        return name()
    }

    companion object {
        private val CONTAINER = Registry.createContainer<StatisticType?>(
            Registry.Resource.STATISTICS
        ) { namespace: String?, properties: Registry.Properties ->
            StatisticTypeImpl(
                NamespaceID.from(
                    namespace!!
                ), properties.getInt("id")
            )
        }

        operator fun get(namespace: String): StatisticType? {
            return CONTAINER[namespace]
        }

        fun getSafe(namespace: String): StatisticType? {
            return CONTAINER.getSafe(namespace)
        }

        fun getId(id: Int): StatisticType? {
            return CONTAINER.getId(id)
        }

        fun values(): Collection<StatisticType?> {
            return CONTAINER.values()
        }
    }
}